crates/wasi-common/src/sys/unix/bsd/hostcalls_impl.rs

```diff
@@ -296,3 +203,85 @@ pub(crate) fn fd_advise(
 
     Ok(())
 }
+
+pub(crate) fn fd_readdir<'a>(
+    os_file: &'a mut OsFile,
+    cookie: wasi::__wasi_dircookie_t,
+) -> Result<impl Iterator<Item = Result<Dirent>> + 'a> {
+    use std::sync::Mutex;
+
+    let dir = match os_file.dir {
+        Some(ref mut dir) => dir,
+        None => {
+            // We need to duplicate the fd, because `opendir(3)`:
+            //     Upon successful return from fdopendir(), the file descriptor is under
+            //     control of the system, and if any attempt is made to close the file
+            //     descriptor, or to modify the state of the associated description other
+            //     than by means of closedir(), readdir(), readdir_r(), or rewinddir(),
+            //     the behaviour is undefined.
+            let fd = (*os_file).try_clone()?;
+            let dir = Dir::from(fd)?;
+            os_file.dir.get_or_insert(Mutex::new(dir))
+        }
+    };
+
+    // Seek if needed. Unless cookie is wasi::__WASI_DIRCOOKIE_START,
+    // new items may not be returned to the caller.
+    if cookie == wasi::__WASI_DIRCOOKIE_START {
+        log::trace!("     | fd_readdir: doing rewinddir");
+        dir.lock().unwrap().rewind();
+    } else {
+        log::trace!("     | fd_readdir: doing seekdir to {}", cookie);
+        let loc = unsafe { SeekLoc::from_raw(cookie as i64) };
+        dir.lock().unwrap().seek(loc);
+    }
+
+    Ok(DirIter(dir.lock().unwrap()).map(|entry| {
+        let (entry, loc): (Entry, SeekLoc) = entry?;
+        Ok(Dirent {
+            name: entry
+                // TODO can we reuse path_from_host for CStr?
+                .file_name()
+                .to_str()?
+                .to_owned(),
+            ino: entry.ino(),
+            ftype: entry.file_type().into(),
+            // Set cookie manually:
+            // * on macOS d_seekoff is not set for some reason
+            // * on FreeBSD d_seekoff doesn't exist; there is d_off but it is
+            //   not equivalent to the value read from telldir call
+            cookie: loc.to_raw().try_into()?,
+        })
+    }))
+}
+
+struct DirIter<'a>(MutexGuard<'a, Dir>);
+
+impl<'a> Iterator for DirIter<'a> {
+    type Item = nix::Result<(Entry, SeekLoc)>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        use libc::readdir;
+        use nix::{errno::Errno, Error};
+
+        unsafe {
+            let errno = Errno::last();
+            let ent = readdir((self.0).0.as_ptr());
+            if ent.is_null() {
+                if errno != Errno::last() {
+                    // According to POSIX man (for Linux though!), there was an error
+                    // if the errno value has changed at some point during the sequence
+                    // of readdir calls
```

Comment:
Linux man says something different:
> To distinguish end of stream and from an error, set errno to zero before calling readdir() and then check the value of errno if NULL is returned.

POSIX says the same:
> Applications wishing to check for error situations should set errno to 0 before calling readdir().  If errno is set to non-zero on return, an error occurred.

Reply:
Interesting, since this [reference] says something different:

> If the end of the directory stream is reached, NULL is returned and errno is not changed. If an error occurs, NULL is returned and errno is set appropriately.

[reference]: https://linux.die.net/man/3/readdir

Also, please note that this bit of code was introduced in 4982878 as a fix to a subtle bug that only emerged when testing the crate in release.

Reply:
The version I quoted is from
> This  page  is  part  of release 5.03 of the Linux man-pages project.

and

> This is POSIX.1-2008 with the 2013 Technical Corrigendum 1 applied.

respectively.

Reply:
Hmm, interesting. Perhaps this is a BSD-only thing (we won't know for sure until we rewrite `fd_readdir` on Linux using `readdir` rather than `readdir_r`. Anyhow, as a compromise, I can remove the mention of Linux in the comment.